### PR TITLE
Try shorthand

### PR DESCRIPTION
**Pull request by marcusklaas**
This is based off https://github.com/rust-lang-nursery/rustfmt/pull/892.

Addresses https://github.com/rust-lang-nursery/rustfmt/issues/867. I'm not 100% content with the implementation, but would like to get some feedback on the output before refining it.

r? @kamalmarhubi, @nrc?

### CODE REVIEW

File: src/expr.rs [hunk 1]
```diff
@@ -148,6 +148,7 @@ impl Rewrite for ast::Expr {
             ast::ExprKind::Closure(capture, ref fn_decl, ref body) => {
                 rewrite_closure(capture, fn_decl, body, self.span, context, width, offset)
             }
+            // ast::ExprKind::Try(..) |
```
Comment (kamalmarhubi) [inline src/expr.rs]: is this meant to be commented out?

File: src/expr.rs [hunk 2]
```diff
@@ -176,48 +177,45 @@ impl Rewrite for ast::Expr {
...
-                             try_opt!(lhs.rewrite(context,
-                                                  try_opt!(width.checked_sub(2)),
-                                                  offset))))
-            }
-            ast::ExprKind::Range(None, None, _range_limits) => {
-                if width >= 2 {
-                    Some("..".into())
-                } else {
-                    None
+            ast::ExprKind::Range(ref lhs, ref rhs, limits) => {
```
Comment (kamalmarhubi) [inline src/expr.rs]: does this belong in this PR?

File: tests/target/expr.rs [hunk 3]
```diff
@@ -267,3 +267,16 @@ fn issue767() {
...
+    let x = aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa..bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb;
+    let y =
+        aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa...bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb;
+    let z = ...x;
+    let infi_range_2 = ...;
+
+    a...b
+
+    // the expr below won't compile for some reason...
+    // let a = 0 ... ;
```
Comment (kamalmarhubi) [inline tests/target/expr.rs]: it's not meant to.. it wouldn't make sense to include infinity :-)

File: src/expr.rs (hunk 2 above)
Comment (kamalmarhubi) [inline src/expr.rs]: ah never mind, didn't catch the note that I should look at the other PR first.

File: tests/target/chains.rs [hunk 4]
```diff
@@ -138,3 +138,16 @@ fn issue587() {
...
+    let y = expr.kaas()?.test();
+    let loooooooooooooooooooooooooooooooooooooooooong = does_this?
+                                                            .look?
+                                                            .good?
+                                                            .should_we_break?
+                                                            .after_the_first_question_mark?;
+    let yyyy = expr?.another?.another?.another?.another?.another?.another?.another?.another?.test();
+    let zzzz = expr?.another?.another?.another?.another?;
+    let aaa = x??????????????????????????????????????????????????????????????????????????;
+}
```
Comment (kamalmarhubi) [inline tests/target/chains.rs]: output looks ok to me!

**Comment by nrc** [comment]
looks good to me. I think the output looks good. Re the FIXME, I think we should never separate the `?` from the subject, i.e., we can try to break the subject, or we can fail, but we shouldn't insert a newline before the `?`.

**Comment by erikjohnston** [comment]
I was trying this and got a panic when using it as the last expression in a block, e.g.:

```rust
fn test() {
    a?
}
```

```
Request to format inverted span: Loc { file: FileMap(tests/source/chains.rs), line: 127, col: CharPos(1) } to Loc { file: FileMap(tests/source/chains.rs), line: 127, col: CharPos(0) }
```

(where line 127 is the last line in the example)

**Comment by marcusklaas** [comment]
Thanks for the test case @erikjohnston, I'll see if I can fix it!

**Comment by marcusklaas** [comment]
Turned out to be a rustc bug. This PR will have to wait until https://github.com/rust-lang/rust/pull/32711 lands and syntax_syntex is updated.

**Comment by marcusklaas** [comment]
The Rust PR (finally) landed. Shouldn't be too long until a syntex update.

**Comment by marcusklaas** [comment]
Updated. we now also convert uses of `try!` to the shorthand `?` when the relevant option is set. This may be the first case of AST manipulation in rustfmt!

Tests won't pass yet -- still need syntex update.

**Comment by nrc** [comment]
@marcusklaas hey, are the changes in syntex now? Would be great to land this, I'd like to release 0.5 soon, and it would be great for this to be part of it (need it for formatting stuff in the Rust repo, amongst other things).

**Comment by kamalmarhubi** [comment]
I just made #971 to bump to syntex 0.31. Not sure if it includes the change but just in case. :-)

**Comment by marcusklaas** [comment]
I don't think it does - but the next version should.. Just a bit longer!

**Comment by kamalmarhubi** [comment]
@marcusklaas have a syntex_syntax bump: https://github.com/rust-lang-nursery/rustfmt/pull/979 :-)

**Comment by marcusklaas** [comment]
Thanks for the syntex update @kamalmarhubi, I'll see if I can rebase this :-)

**Comment by marcusklaas** [comment]
Yo @nrc, this should be ready to merge now. I've also implemented a break of assignments when it's more compact. It should mitigate the need for tab indentation in chains; I'll look into that a bit more.

**Comment by marcusklaas** [comment]
Closes https://github.com/rust-lang-nursery/rustfmt/issues/985.

**Comment by marcusklaas** [comment]
Supersedes https://github.com/rust-lang-nursery/rustfmt/pull/751 too.

**Comment by marcusklaas** [comment]
Addresses https://github.com/rust-lang-nursery/rustfmt/issues/574 as well.

File: src/chains.rs [hunk 5]
```diff
@@ -184,12 +191,27 @@ pub fn rewrite_chain(expr: &ast::Expr,
...
+fn join_rewrites(rewrites: &[String], subexps: &[&ast::Expr], connector: &str) -> String {
+    let mut rewrite_iter = rewrites.iter();
+    let mut result = rewrite_iter.next().unwrap().clone();
+
+    for (rewrite, expr) in rewrite_iter.zip(subexps.iter()) {
+        match expr.node {
+            ast::ExprKind::Try(_) => (),
+            _ => result.push_str(connector),
+        };
+        result.push_str(&rewrite[..]);
```
Comment (nrc) [inline src/chains.rs]: nit: probably don't need the `[..]`

**Pull request merged**